app/jobs/unzip_job.rb

```diff
@@ -0,0 +1,53 @@
+# frozen_string_literal: true
+
+require 'zip'
+
+# Change work version from a zipfile type to a browser type by unzipping the attached zip file.
+class UnzipJob < BaseDepositJob
+  queue_as :default
+
+  def perform(work_version)
+    zip_attached_file = work_version.attached_files.last
+    destroy_attached_files_except(zip_attached_file, work_version)
+    Zip::File.open(filepath_for(zip_attached_file)) do |zip_file|
+      Dir.mktmpdir do |temp_dir|
+        zip_file.each do |entry|
+          next if ignore?(entry)
+
+          attach_file(entry:, temp_dir:, work_version:)
+        end
+      end
+    end
+    zip_attached_file.destroy
+    work_version.upload_type = 'browser'
+    work_version.unzip_complete!
+  end
+
+  private
+
+  def destroy_attached_files_except(except_attached_file, work_version)
+    work_version.attached_files.each do |attached_file|
+      work_version.attached_files.destroy(attached_file) unless attached_file == except_attached_file
+    end
+  end
+
+  def ignore?(entry)
+    !entry.file? || entry.name.start_with?('__MACOSX') || entry.name.end_with?('.DS_Store')
+  end
+
+  def filepath_for(attached_file)
+    ActiveStorage::Blob.service.path_for(attached_file.blob.key)
+  end
```

Comment:
This could also be a method worth considering for the `AttachedFile` model so it's responsible for knowing its own file path.

Reply:
I just went to do this. However, AttachedFile is a bit of a confusing mess because it might have different kinds of Blobs. (Some of the existing methods only work with certain kinds of Blobs, which is part of the problem.) This method will only work for ActiveStorage Blobs.

So I'm inclined to leave this closer to its context then add to AttachedFile where it can generate confusion.

Reply:
Sounds good. Thanks for looking into it.